bazel/multitool/extension.bzl

```diff
@@ -0,0 +1,280 @@
+           tool.bzl       (scaffolding for the tool target and toolchain declarations in toolchains/BUILD.bazel)
+       toolchain_info.bzl (common scaffolding for toolchain declarations)
+       tools.bzl          (a file containing one utility method, "register_tools", to support WORKSPACE users)
+
+       (additional BUILD.bazel and a WORKSPACE file are included as required by Bazel)
+
+To keep things orderly, we keep all the toolchain Bazel goo in the [hub] repo and only stash
+the binaries in the [hub].[tool].[os]_[cpu] repos. It's a conscious decision not to place some
+fragments of the toolchain definitions in the latter repos to make the dependencies run exactly
+one way:
+    [hub] -> [hub].[tool].[os]_[cpu].
+
+This implementation depends on rendering a number of templates, which are defined in sibling
+folders and managed by the templates starlark file.
+
+Note that we intend to support both bzlmod and non-bzlmod setups, so `hub` intentionally avoids
+a register_toolchains call.
+"""
+
+load("//bazel/multitool/private:lockfile.bzl", "lockfile")
+load("//bazel/multitool/private:templates.bzl", "templates")
+
+_DEFAULT_HUB_NAME = "multitool"
+
+def file_extension(os):
+    if os == "windows":
+        return ".exe"
+    return ""
+
+def _download_extract_tool(rctx, tool_name, binary):
+    target_filename = "{os}_{cpu}_executable{ext}".format(
+        cpu = binary["cpu"],
+        os = binary["os"],
+        ext = file_extension(binary["os"]),
+    )
+    target_executable = "tools/{tool_name}/{filename}".format(
+        tool_name = tool_name,
+        filename = target_filename,
+    )
+
+    if binary["kind"] == "file":
+        rctx.download(
+            url = binary["url"],
+            sha256 = binary["sha256"],
+            output = target_executable,
+            executable = True,
+            headers = binary.get("headers", {})
+        )
+    elif binary["kind"] == "archive":
+        archive_path = "tools/{tool_name}/{os}_{cpu}_archive".format(
+            tool_name = tool_name,
+            cpu = binary["cpu"],
+            os = binary["os"],
+        )
+
+        rctx.download_and_extract(
+            url = binary["url"],
+            sha256 = binary["sha256"],
+            output = archive_path,
+            type = binary.get("type", ""),
+            headers = binary.get("headers", {})
+        )
+
+        # link to the executable
+        archive_file = "{archive_path}/{file}".format(archive_path = archive_path, file = binary["file"])
+        if not rctx.path(archive_file).exists:
+            fail("{tool_name} ({os}, {cpu}): Cannot find {file} in archive from {url}".format(
+                tool_name = tool_name,
+                os = binary["os"],
+                cpu = binary["cpu"],
+                file = archive_file,
+                url = binary["url"],
+            ))
+        rctx.symlink(archive_file, target_executable)
```

Comment:
What does it do in case of Windows?

Reply:
Great question. IIUC, the only practical way to use bazel on windows is to enable symlinks.

Reply:
Either way.... symlinks is a general bazel issue, separate from this PR.